File: Data Structures & Algorithms/decode-ways/submission-2.cpp
class Solution
{
public:
    int numDecodings(string s)
    {
        int length = s.length();

        vector<int> num(length);
        
        if (s[0] == '0')
        {
            return 0;
        }

        num[0] = 1;
        
        if (s[0] == '1' || s[0] == '2')
        {
            int prev = s[0] - '0';
            int curr = s[1] - '0';
            if (curr == 0 || prev * 10 + curr > 26)
            {
                num[1] = 1;
            }
            else
            {
                num[1] = 2;
            }
        }
        else
        {
            if (s[1] == '0')
            {
                return 0;
            }
            num[1] = 1;
        }
        for (int i = 2; i < length; i++)
        {
            int prev = s[i - 1] - '0';
            int curr = s[i] - '0';
            if (prev == 1 || prev == 2)
            {
                if (curr == 0)
                {
                    num[i] = num[i - 2];
                }
                else if (prev * 10 + curr > 26) {
                    num[i] = num[i - 1];
                }
                else
                {
                    num[i] = num[i - 1] + num[i - 2];
                }
            }
            else
            {
                if (curr == 0)
                {
                    return 0;
                }
                num[i] = num[i - 1];
            }
        }
        return num[length - 1];
    }
};

File: Data Structures & Algorithms/count-connected-components/submission-2.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        visited = set()
        def bfs(node):
            q = deque()
            q.append(node)
            visited.add(node)
            while q:
                curr = q.popleft()
                for entry in adj[curr]:
                    if entry not in visited:
                        q.append(entry)
                        visited.add(entry)
        
        res = 0
        for i in range(n):
            if i not in visited:
                bfs(i)
                res += 1
        return res